open_spiel/games/einstein_wurfelt_nicht/einstein_wurfelt_nicht.cc

```diff
@@ -0,0 +1,557 @@
+                         GameType::ChanceMode::kExplicitStochastic,
+                         GameType::Information::kPerfectInformation,
+                         GameType::Utility::kZeroSum,
+                         GameType::RewardModel::kTerminal,
+                         /*max_num_players=*/2,
+                         /*min_num_players=*/2,
+                         /*provides_information_state_string=*/false,
+                         /*provides_information_state_tensor=*/false,
+                         /*provides_observation_string=*/true,
+                         /*provides_observation_tensor=*/true,
+                         /*parameter_specification=*/{}  // no parameters
+                         };
+
+std::shared_ptr<const Game> Factory(const GameParameters& params) {
+  return std::shared_ptr<const Game>(new EinsteinWurfeltNichtGame(params));
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+RegisterSingleTensorObserver single_tensor(kGameType.short_name);
+
+Color PlayerToColor(Player player) {
+  SPIEL_CHECK_NE(player, kInvalidPlayer);
+  return static_cast<Color>(player);
+}
+
+Player ColorToPlayer(Color color) {
+  switch (color) {
+    case Color::kBlack:
+      return kBlackPlayerId;
+    case Color::kWhite:
+      return kWhitePlayerId;
+    default:
+      SpielFatalError("No player for this color");
+  }
+}
+
+Color OpponentColor(Player player) {
+  Color player_color = PlayerToColor(player);
+  if (player_color == Color::kBlack) {
+    return Color::kWhite;
+  } else if (player_color == Color::kWhite) {
+    return Color::kBlack;
+  } else {
+    SpielFatalError("Player should be either black or white");
+  }
+}
+
+std::vector<std::vector<int>> GetAllPermutations() {
+  std::vector<std::vector<int>> all_permutations;
+  std::vector<int> nums = {1, 2, 3, 4, 5, 6};
+
+  do {
+    all_permutations.push_back(nums);
+  } while (std::next_permutation(nums.begin(), nums.end()));
+
+  return all_permutations;
+}
+
+std::string CoordinatesToDirection(int row, int col) {
+  std::string direction;
+  if (row == col) {
+    direction = "diag";
+  } else if (row == -1) {
+    direction = "up";
+  } else if (row == 1) {
+    direction = "down";
+  } else if (col == 1) {
+    direction = "right";
+  } else if (col == -1) {
+    direction = "left";
+  } else {
+    std::cout << "r2: " << row << "c2: " << col << std::endl;
+    SpielFatalError("Unrecognized cube's movement");
+  }
+  return direction;
+}
+
+}  // namespace
+
+EinsteinWurfeltNichtState::EinsteinWurfeltNichtState(
+    std::shared_ptr<const Game> game, int rows, int cols)
+    : State(game),
+      rows_(rows),
+      cols_(cols),
+      turns_(-1),
+      cur_player_(kChancePlayerId),
+      prev_player_(kBlackPlayerId) {
+  SPIEL_CHECK_GT(rows_, 1);
+  SPIEL_CHECK_GT(cols_, 1);
+  board_.fill(Cube{Color::kEmpty, -1});
+
+  winner_ = kInvalidPlayer;
+  cubes_[0] = cubes_[1] = kNumPlayerCubes;
+}
+
+void EinsteinWurfeltNichtState::SetupInitialBoard(
```

Comment:
This function generates all 6! permutations and takes factorial space. In this case it's a vector of 720 vectors.

I think we should change this to one of two options:

1. There are known constant-space unranking functions for permutations.  (See this: https://stackoverflow.com/questions/31702845/permutations-unrank)

2. Instead of a single chance node with all the permutations, you can use 12 chances nodes (each with <= 6 outcomes). The chance outcomes labeled 0-5 correspond to the die that is placed in the next position. At first, all of the outcomes are "legal". Next, only the 5 remaining are legal. Next, only the 4 remaining are legal.. etc.

I don't have a strong preference. (2) is a bit less cryptic and consistent with other games, and (1) makes the code a bit more cryptic, but is completely fine as long as it's correct. If you chose 1, I would add a test to make sure the unrank(p) is in the set of permutations and that it uniquely generates all of them if you loop from 0 to 6! - 1.

Reply:
Hi @giogix2, 

I completely forgot that we already have the permutation unrank in the code base! ... https://github.com/google-deepmind/open_spiel/blob/b8225eb95a357bd9ab13d098e5e3cddcc4698ff1/open_spiel/utils/combinatorics.h#L139

So let's go with that one (it has been tested independently, so you can assume it works).

Reply:
> Hi @giogix2,
> 
> I just remembered that we already have the permutation unrank in the code base! ...
> 
> https://github.com/google-deepmind/open_spiel/blob/b8225eb95a357bd9ab13d098e5e3cddcc4698ff1/open_spiel/utils/combinatorics.h#L139
> 
> So let's go with that one (it has been tested independently, so you can assume it works).

Oh nice. This helps. I should have asked earlier about this. I missed that in the utils.

Looking into the combinatorics utils I noticed that there is an implementation of Factorial https://github.com/google-deepmind/open_spiel/blob/b8225eb95a357bd9ab13d098e5e3cddcc4698ff1/open_spiel/utils/combinatorics.cc#L56, which is used in pathfinding. I'm wondering if we also should remove this hardcoded value for the permutations https://github.com/giogix2/open_spiel/blob/einstein_wurfelt_nicht/open_spiel/games/einstein_wurfelt_nicht/einstein_wurfelt_nicht.h#L48.

I'll apply the suggestions in the next few days (likely the weekend).

Reply:
I think it's fine to leave this hard-coded since it's a constant and the game doesn't support different board sizes.

Reply:
Done. I have replaced the logic with UnrankPermutation().